micro_http/src/server.rs

```diff
@@ -0,0 +1,834 @@
+    pub fn process<F>(&self, callable: F) -> ServerResponse
+    where
+        F: Fn(&Request) -> Response,
+    {
+        let http_response = callable(self.inner());
+        ServerResponse::new(http_response, self.id)
+    }
+}
+
+/// Wrapper over `Response` which adds an identification token.
+pub struct ServerResponse {
+    /// Inner response.
+    response: Response,
+    /// Identification token.
+    id: u64,
+}
+
+impl ServerResponse {
+    fn new(response: Response, id: u64) -> ServerResponse {
+        ServerResponse { response, id }
+    }
+}
+
+/// Describes the state of the connection as far as data exchange
+/// on the stream is concerned.
+#[derive(PartialOrd, PartialEq)]
+enum ClientConnectionState {
+    AwaitingIncoming,
+    AwaitingOutgoing,
+    Closed,
+}
+
+/// Wrapper over `HttpConnection` which keeps track of yielded
+/// requests and absorbed responses.
+struct ClientConnection<T> {
+    /// The `HttpConnection` object which handles data exchange.
+    connection: HttpConnection<T>,
+    /// The state of the connection in the `epoll` structure.
+    state: ClientConnectionState,
+    /// Represents the difference between yielded requests and
+    /// absorbed responses.
+    /// This has to be `0` if we want to drop the connection.
+    in_flight_response_count: u32,
+}
+
+impl<T: Read + Write> ClientConnection<T> {
+    fn new(connection: HttpConnection<T>) -> Self {
+        ClientConnection {
+            connection,
+            state: ClientConnectionState::AwaitingIncoming,
+            in_flight_response_count: 0,
+        }
+    }
+
+    fn read(&mut self) -> Result<Vec<Request>> {
+        // Data came into the connection.
+        let mut parsed_requests = vec![];
+        match self.connection.try_read() {
+            Err(ConnectionError::ConnectionClosed) => {
+                // Connection timeout.
+                self.state = ClientConnectionState::Closed;
+                // We don't want to propagate this to the server and we will
+                // return no requests and wait for the connection to become
+                // safe to drop.
+                return Ok(vec![]);
+            }
+            Err(ConnectionError::StreamError(inner)) => {
+                // Reading from the connection failed.
+                // We should try to write an error message regardless.
+                let mut internal_error_response =
+                    Response::new(Version::Http11, StatusCode::InternalServerError);
+                internal_error_response.set_body(Body::new(inner.to_string()));
+                self.connection.enqueue_response(internal_error_response);
+            }
+            Err(ConnectionError::ParseError(inner)) => {
+                // An error occurred while parsing the read bytes.
+                // Check if there are any valid parsed requests in the queue.
+                while let Some(_discarded_request) = self.connection.pop_parsed_request() {}
```

Comment:
We might want to remove this line of code in the future, our server could be more lenient and accept valid previously parsed requests, and not drop all of them when one is bad.

The consequence that needs to be researched is that responses for the burst of requests would then come out-of-order (1st response will be the parse error for the last request, followed by the rest of responses for the valid requests in the right order).